signbank/dictionary/update.py

```diff
@@ -764,6 +774,316 @@ def confirm_import_gloss_csv(request):
+        csv_form.fields["dataset"].queryset = csv_form.fields["dataset"].queryset.filter(
+            id__in=[x.id for x in allowed_datasets])
+        return render(request, "dictionary/import_nzsl_share_gloss_csv.html",
+                      {'import_csv_form': csv_form}, )
+
+
+@login_required
+@permission_required('dictionary.import_csv')
+@transaction.atomic()
+def confirm_import_nzsl_share_gloss_csv(request):
+    """This view adds the data to database if the user confirms the action"""
+    if request.method == 'POST':
+        if 'cancel' in request.POST:
+            # If user cancels adding data, flush session variables
+            if 'dataset_id' in request.session: del request.session['dataset_id']
+            if 'glosses_new' in request.session: del request.session['glosses_new']
+            # Set a message to be shown so that the user knows what is going on.
+            messages.add_message(request, messages.WARNING, _('Cancelled adding CSV data.'))
+            return HttpResponseRedirect(reverse('dictionary:import_nzsl_share_gloss_csv'))
+
+        elif 'confirm' in request.POST:
+            glosses_added = []
+            dataset = None
+            translations = []
+            comments = []
+            videos = []
+            new_glosses = {}
+            bulk_create_gloss = []
+            bulk_update_glosses = []
+            bulk_semantic_fields = []
+
+            if 'glosses_new' and 'dataset_id' in request.session:
+                dataset = Dataset.objects.get(id=request.session['dataset_id'])
+                language_en = Language.objects.get(name="English")
+                language_mi = Language.objects.get(name="Māori")
+                gloss_content_type = ContentType.objects.get_for_model(Gloss)
+                video_type = FieldChoice.objects.get(field="video_type", english_name="validation")
+                site = Site.objects.get_current()
+                comment_submit_date = datetime.datetime.now(tz=get_current_timezone())
+                semantic_fields = FieldChoice.objects.filter(
+                    field="semantic_field"
+                ).values_list("english_name", "pk")
+                semantic_fields_dict = {field[0]: field[1] for field in semantic_fields}
+
+                for row_num, gloss in enumerate(request.session['glosses_new']):
+                    # will iterate over these glosses again after bulk creating and to make sure
+                    # we preserve order
+                    new_glosses[str(row_num)] = gloss
+                    bulk_create_gloss.append(Gloss(
+                        dataset=dataset,
+                        # need to make idgloss unique in dataset,
+                        # but gloss word can appear in multiple rows, so
+                        # idgloss will be updated to word:pk in second step
+                        idgloss=f"{gloss['word']}_row{row_num}",
+                        idgloss_mi=gloss.get("maori", None),
```

Comment:
**question**: is it not possible to also have the maori gloss word appear in multiple rows?

Reply:
In theory yes. However there is a unique constraint on (dataset, idgloss) which is why I only added it to the idgloss